### PR TITLE
Add userspace and personal keymaps

### PR DESCRIPTION
**Pull request by mguterl**
## Description

This PR adds my personal keymaps and userspace.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/planck/keymaps/mguterl/readme.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+At a very high level I use 4 layers:
+
+* Qwerty - Letters
+* Util   - Vimkeys for navigation and other convenient things
+* Raise  - Numbers and F keys
+* Lower  - Symbols
+
+```sh
+make planck/rev5:mguterl:dfu         # For Planck rev5 or earlier and Planck Light
+make planck/rev6:mguterl:dfu-util    # For Planck rev6
```
Comment (drashna) [inline keyboards/planck/keymaps/mguterl/readme.md]: Actually, we have `:flash` now.  This handles the detection and will attempt to flash using the "correct" method based on the controller and bootloader set.

File: keyboards/keebio/levinson/keymaps/mguterl/readme.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+At a very high level I use 4 layers:
+
+* Qwerty - Letters
+* Util   - Vimkeys for navigation and other convenient things
+* Raise  - Numbers and F keys
+* Lower  - Symbols
+
+```sh
+make keebio/levinson/rev2:mguterl:dfu-util
```
Comment (drashna) [inline keyboards/keebio/levinson/keymaps/mguterl/readme.md]:
```suggestion
make keebio/levinson/rev2:mguterl:flash
```

File: keyboards/foxlab/leaf60/universal/keymaps/mguterl/readme.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Michael Guterl's Leaf60 WKL Layout
+
+Check out my [userspace](/users/mguterl) for the custom keycodes that are used
+in keymap.c.
+
+This is a pretty standard 60% WKL layout. I add a couple of layers to add
+some conveniences and add vimkeys for navigation.
+
+```sh
+make foxlab/leaf60/universal:mguterl:dfu
```
Comment (drashna) [inline keyboards/foxlab/leaf60/universal/keymaps/mguterl/readme.md]:
```suggestion
make foxlab/leaf60/universal:mguterl:flash
```

File: keyboards/preonic/keymaps/mguterl/readme.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+At a very high level I use 4 layers:
+
+* Qwerty - Letters
+* Util   - Vimkeys for navigation and other convenient things
+* Raise  - Numbers and F keys
+* Lower  - Symbols
+
+```sh
+make preonic/rev3:mguterl:dfu-util
```
Comment (drashna) [inline keyboards/preonic/keymaps/mguterl/readme.md]:
```suggestion
make preonic/rev3:mguterl:flash
```

File: keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/readme.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+This is a pretty standard HHKB layout. I add a couple of layers to add
+some conveniences and add vimkeys for navigation.
+
+If you flash a layout without RESET then you can unplug the keyboard, hold
+spacebar and escape, and then plug the keyboard back. This sequence of
+actions will allow you to flash the keyboard.
+
+```sh
+make wilba_tech/rama_works_m60a:mguterl:dfu-util
```
Comment (drashna) [inline keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/readme.md]:
```suggestion
make wilba_tech/rama_works_m60a:mguterl:flash
```

File: keyboards/planck/keymaps/mguterl/readme.md (hunk 1 above)
Comment (mguterl) [reply keyboards/planck/keymaps/mguterl/readme.md]: This is awesome, thanks for the tip!

**Comment by mguterl** [comment]
> Also, just a heads up, you could move the levigson and planck keymaps to `/layouts/community/ortho_4x12/` instead. It would use the same keymap for both, and may be easier to maintain.
> 
> The same may apply to your 60% boards, too.

@drashna I'm definitely interested in doing this for my 4x12 boards. How do I flash if I'm using one of these community layouts?

File: keyboards/keebio/levinson/keymaps/mguterl/keymap.c [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+ * | Shift|   Z  |   X  |   C  |   V  |   B  |   N  |   M  |   ,  |   .  |   /  |Enter |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * | GESC | Ctrl | Alt  | GUI  |Lower |SpcUtl|SpcUtl|Raise | Left | Down |  Up  |Right |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_QWERTY] = LAYOUT_ortho_4x12(
+    HPR_TAB, KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC, \
+    ESC_CTL, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, SFT_ENT , \
+    KC_GESC, KC_LCTL, KC_LALT, KC_LGUI, LOWER,   SPC_UTL, SPC_UTL, RAISE,   KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT \
```
Comment (fauxpark) [inline keyboards/keebio/levinson/keymaps/mguterl/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+          return false;
+          break;
+        case BACKLIT:
+          if (record->event.pressed) {
+            register_code(KC_RSFT);
+            #ifdef BACKLIGHT_ENABLE
+              backlight_step();
+            #endif
+            #ifdef __AVR__
+            PORTE &= ~(1<<6);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
            writePinLow(E6);
```

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            #endif
+          } else {
+            unregister_code(KC_RSFT);
+            #ifdef __AVR__
+            PORTE |= (1<<6);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
            writePinHigh(E6);
```

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    } else {
+      if (clockwise) {
+        muse_tempo+=1;
+      } else {
+        muse_tempo-=1;
+      }
+    }
+  } else {
+    if (clockwise) {
+      register_code(KC_PGDN);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
      tap_code(KC_PGDN);
```

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+      unregister_code(KC_PGDN);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
```

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    } else {
+      register_code(KC_PGUP);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
      tap_code(KC_PGUP);
```

File: keyboards/preonic/keymaps/mguterl/keymap.c [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+      unregister_code(KC_PGUP);
```
Comment (fauxpark) [inline keyboards/preonic/keymaps/mguterl/keymap.c]:
```suggestion
```

File: keyboards/planck/keymaps/mguterl/keymap.c [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+      return false;
+      break;
+    case BACKLIT:
+      if (record->event.pressed) {
+        register_code(KC_RSFT);
+        #ifdef BACKLIGHT_ENABLE
+          backlight_step();
+        #endif
+        #ifdef KEYBOARD_planck_rev5
+          PORTE &= ~(1<<6);
```
Comment (fauxpark) [inline keyboards/planck/keymaps/mguterl/keymap.c]:
```suggestion
          writePinLow(E6);
```

File: keyboards/planck/keymaps/mguterl/keymap.c [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+        #endif
+      } else {
+        unregister_code(KC_RSFT);
+        #ifdef KEYBOARD_planck_rev5
+          PORTE |= (1<<6);
```
Comment (fauxpark) [inline keyboards/planck/keymaps/mguterl/keymap.c]:
```suggestion
          writePinHigh(E6);
```

File: keyboards/keebio/levinson/keymaps/mguterl/keymap.c [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+      return false;
+      break;
+    case BACKLIT:
+      if (record->event.pressed) {
+        register_code(KC_RSFT);
+        #ifdef BACKLIGHT_ENABLE
+          backlight_step();
+        #endif
+        #ifdef KEYBOARD_planck_rev5
+          PORTE &= ~(1<<6);
```
Comment (fauxpark) [inline keyboards/keebio/levinson/keymaps/mguterl/keymap.c]:
```suggestion
          writePinLow(E6);
```

File: keyboards/keebio/levinson/keymaps/mguterl/keymap.c [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+        #endif
+      } else {
+        unregister_code(KC_RSFT);
+        #ifdef KEYBOARD_planck_rev5
+          PORTE |= (1<<6);
```
Comment (fauxpark) [inline keyboards/keebio/levinson/keymaps/mguterl/keymap.c]:
```suggestion
          writePinHigh(E6);
```

File: keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/keymap.c [hunk 17]
```diff
@@ -0,0 +1,83 @@
+#include QMK_KEYBOARD_H
+
+#define _______ KC_TRNS
```
Comment (fauxpark) [inline keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/keymap.c]:
```suggestion
```

QMK defines this already.

File: keyboards/keebio/levinson/keymaps/mguterl/keymap.c (hunk 6 above)
Comment (mguterl) [reply keyboards/keebio/levinson/keymaps/mguterl/keymap.c]: Thanks!

**Comment by fauxpark** [comment]
Yep. Specifically, the `writePinLow/High()` macros have implementations for both ARM and AVR, though it doesn't matter too much here because I think the Planck rev6 and Preonic rev3 don't have this LED (and it would be on a different pin). It's just for consistency.

And yeah, I opened a PR just now to fix this in the default keymaps, so we won't have to keep reviewing them in future, thanks for reminding me 😄 #7098

**Comment by mguterl** [comment]
> > Also, just a heads up, you could move the levigson and planck keymaps to `/layouts/community/ortho_4x12/` instead. It would use the same keymap for both, and may be easier to maintain.
> > The same may apply to your 60% boards, too.
> 
> @drashna I'm definitely interested in doing this for my 4x12 boards. How do I flash if I'm using one of these community layouts?

@drashna I dug in a bit more after finding the documentation on layouts and I think I have a good direction now. I will fully test my changes on my Levinson tomorrow and then I'll push the changes up for review.

**Comment by drashna** [comment]
> > > Also, just a heads up, you could move the levigson and planck keymaps to `/layouts/community/ortho_4x12/` instead. It would use the same keymap for both, and may be easier to maintain.
> > > The same may apply to your 60% boards, too.
> > 
> > 
> > @drashna I'm definitely interested in doing this for my 4x12 boards. How do I flash if I'm using one of these community layouts?
> 
> @drashna I dug in a bit more after finding the documentation on layouts and I think I have a good direction now. I will fully test my changes on my Levinson tomorrow and then I'll push the changes up for review.

Awesome!

And as for flashing, just like normal.  QMK Firmware's make files take care of the detection and everything.  

For instance, I have my own ortho_4x12 folder in the layouts folder for this.  I can run:
```
make planck/rev6:drashna
make planck/ez:drashna
make planck/light:drashna
make lets_split:drashna
make vitamins_included:drashna
make levinson:drashna
make contra:drashna
```

Etc.  All of these will use the exact same file to compile.   And there are ways to make sure that only specific boards have specific code enabled, etc.  

For more details: https://docs.qmk.fm/#/feature_layouts

**Comment by mguterl** [comment]
@drashna Thanks! I didn't realize how the hierarchy worked and that the planck / levinson targets would lookup ortho 4x12 community layouts. I really love how userspace and layouts work, both features seem to be very well designed.

Can you approve my PR again? I pushed a small readme update after flashing the board at work today.

File: keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/keymap.c [hunk 18]
```diff
@@ -0,0 +1,82 @@
+#include QMK_KEYBOARD_H
+
+
+#define QWER 0
+#define GAMING 1
+#define HHKB 2
+#define UTIL 3
+#define LIGHTING 4
```
Comment (noroadsleft) [inline keyboards/wilba_tech/rama_works_m60_a/keymaps/mguterl/keymap.c]: Lines 4-8 should be replaced with:

```c
enum layer_names {
    QWER,
    GAMING,
    HHKB,
    UTIL,
    LIGHTING,
};
```

This is identical to the set of defines you have here.

**Pull request merged**